path_selection/path_selection.go

```diff
@@ -0,0 +1,275 @@
+// The ID and PublicKey are found in the mix descriptors while the
+// Commands are specified by this function. The recipientID is only
+// used to create forward paths, when isSURB is set to false.
+func (r *RouteFactory) newPathVector(till time.Duration,
+	delays []float64,
+	descriptors []*pki.MixDescriptor,
+	recipientID [constants.RecipientIDLength]byte,
+	isSURB bool) ([]*sphinx.PathHop, *[constants.SURBIDLength]byte, error) {
+
+	path := make([]*sphinx.PathHop, r.nrHops)
+	var surbID *[constants.SURBIDLength]byte = nil
+	keys, err := r.getHopEpochKeys(till, delays, descriptors)
+	if err != nil {
+		return nil, nil, err
+	}
+	for i := range path {
+		path[i] = new(sphinx.PathHop)
+		copy(path[i].ID[:], descriptors[i].ID[:])
+		path[i].PublicKey = keys[i]
+		if i < r.nrHops-1 {
+			// Non-terminal hop, add the delay.
+			delay := new(commands.NodeDelay)
+			delay.Delay = uint32(delays[i])
+			path[i].Commands = append(path[i].Commands, delay)
+		} else {
+			if isSURB {
+				surbReply := new(commands.SURBReply)
+				surbID = &[constants.SURBIDLength]byte{}
+				_, err := rand.Reader.Read(surbID[:])
+				if err != nil {
+					return nil, nil, err
+				}
+				surbReply.ID = *surbID
+				path[i].Commands = append(path[i].Commands, surbReply)
+			} else {
+				// Terminal hop, add the recipient.
+				recipient := new(commands.Recipient)
+				copy(recipient.ID[:], recipientID[:])
+				path[i].Commands = append(path[i].Commands, recipient)
+			}
+		}
+	}
+	return path, surbID, nil
+}
+
+// next returns a new forward path, reply path, SURB_ID and error.
+// This implements section 5.2 Path selection algorithm of the
+// Panoramix Mix Network End-to-end Protocol Specification
+// see https://github.com/Katzenpost/docs/blob/master/specs/end_to_end.txt
+// The generated forward and reply paths are intended to be used
+// with the Poisson Stop and Wait ARQ, an end to end reliable transmission
+// protocol for mix networks using the Poisson mix strategy.
+func (r *RouteFactory) next(senderProviderName, recipientProviderName string, recipientID [constants.RecipientIDLength]byte) ([]*sphinx.PathHop, []*sphinx.PathHop, *[constants.SURBIDLength]byte, time.Duration, error) {
+
+	var rtt time.Duration
+
+	// 1. Sample all forward and SURB delays.
+	forwardDelays := getDelays(r.lambda, r.nrHops)
+	replyDelays := getDelays(r.lambda, r.nrHops)
+
+	// 2. Ensure total delays doesn't exceed (time_till next_epoch) +
+	//    2 * epoch_duration, as keys are only published 3 epochs in
+	//    advance.
+	_, _, till := epochtime.Now()
+	forwardDuration, err := durationFromFloat(sum(forwardDelays))
+	if err != nil {
+		return nil, nil, nil, rtt, err
+	}
+	replyDuration, err := durationFromFloat(sum(replyDelays))
+	if err != nil {
+		return nil, nil, nil, rtt, err
+	}
+
+	rtt = forwardDuration + replyDuration
+
+	if forwardDuration+replyDuration > till+(2*epochtime.Period) {
+		return nil, nil, nil, rtt, errors.New("selected delays exceed permitted epochtime range")
```

Comment:
shouldn't this method resample delays in this case, rather than erring out some known fraction of the time?

Reply:
yes for that particular error we could resample delays